File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
# 백준 집합의 표현 https://www.acmicpc.net/problem/1717

import sys

input = sys.stdin.readline
sys.setrecursionlimit(100000)


def find(x):
  # 1. 종료조건
  # 대표값을 찾을 때까지 (= 나랑 부모랑 같을 때까지)
  if x == parent[x]:
    return parent[x]
  
  # 2. 재귀식
  parent[x] = find(parent[x]) # 경로압축
  # 나랑 위랑 같을 때까지 반복함 
  # x != parent[x]: # 한 칸 올라감. 그럼 x=parent[x]가 됨. 한칸 올라간 parent[x]를 넘겨서 계속 올라감.ㄴ
  return parent[x]

def union(x, y):
  x_root = find(x) # x가 속한 집합
  y_root = find(y) # y가 속한 집합

  if x_root == y_root:
    return
  
  if x_root < y_root:
    parent[y_root] = x_root
  else:
    parent[x_root] = y_root

n, m = map(int, input().split())
parent = list(range(n + 1))

for _ in range(m):
  command, a, b = map(int, input().split())

  if command == 0:
    union(a, b)
  else: # command == 1:
    if find(a) == find(b):
      print("YES")
    else:
      print("NO")

File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.util.Scanner;

public class Main {
	
	static int n;
	static int m;
	static int[] parent;
	
	public static void main(String[] args) {
	
		Scanner sc = new Scanner(System.in);
		
		n = sc.nextInt(); // 0부터 n (정점 수)
		m = sc.nextInt(); // 쿼리 수
		parent = new int[n + 1];
		
		// parent 배열 초기화 (자기 자신을 부모로 설정)
		for(int i = 1; i <= n; i++) {
			parent[i] = i;
		}
		
		// m줄에 걸쳐 입력받기
		// 쿼리를 저장할 필요 없이 바로 처리 가능
		for(int i = 0; i < m; i++) {
			int cmd = sc.nextInt(); // 0 or 1
			int a = sc.nextInt(); 
			int b = sc.nextInt();
			
			// cmd가 0이면 union, 1이면 find
			if(cmd == 0) {
				union(a, b);
			} else {
				// 두 원소가 같은 집합에 속해있다면
				if(find(a) == find(b)) {
					System.out.println("yes");
				} else {
					System.out.println("no");
				}
			}
		}
		
		sc.close();
	}//main
	
	// find: 루트 부모 찾기
	public static int find(int x) {
		if(x != parent[x]) {
			parent[x] = find(parent[x]);
		}
		return parent[x];
	}//
	
	// union: 두 집합 합치기
	public static void union(int x, int y) {
		// 루트 찾기 (x가 속한 집합과 y가 속한 집합을 조회)
		int rootX = find(x);
		int rootY = find(y);
		
		// 루트끼리 연결
		if(rootX != rootY) {
			parent[rootY] = rootX;
		}
	}//
	
}//